<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Examples</title>
</head>
<body>

<script type="text/javascript">
	var examples = {
		'ErrorObjectDef': function(){
		//▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾//
			var error = new Error("Oh, error!");
			console.dir(error);
		//▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴//
		},
		'ErrorObject': function(){
		//▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾//
			function validateEmail(email) {
				return email.match(/[\w.-]@\w+\.\w+/) ?
					email :
					new Error(`invalid email: ${email}`);
			}

			var validEmail = validateEmail('name@domaincom');

			if(validEmail instanceof Error) {
				console.error(`Error: ${validEmail.message}`);
			} else {
				console.log(`Valid email: ${validEmail}`);
			}
		//▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴//
		},
		'trycatch': function(){
		//▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾//
			"use strict";
			try{
				x = 2;
				console.log(`x=${x}`);
			}catch(err){
				console.error(`Error: ${err.message}`);
				console.dir(err);
			}


		//▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴//
		},
		'trycatch2': function(){
		//▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾//
			"use strict";
            function validateEmail(email) {
                return email.match(/[\w.-]@\w+\.\w+/) ?
                    email :
                    new Error(`invalid email: ${email}`);
            }

            var email = 'namedomain.com';
            try {
				const validatedEmail = validateEmail(email);
				if(validatedEmail instanceof Error) {
					console.error(`Anticipated Error: ${validatedEmail.message}`);
				} else {
					console.log(`Valid email: ${validatedEmail}`);
				}
			} catch(err) {
				console.error(`Unanticipated Error: ${err.message}`);
			}
		//▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴//
		},
		'trycatchScope': function(){
		//▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾//
			try{
				x + 2;
			}catch(err){
				console.error(`Error: ${err.message}`);
				console.dir(err); // ok
			}

			console.dir(err); // ReferenceError: err is not defined

		//▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴//
		},
		'trycatchfinally': function(){
		//▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾//
			try{
				console.log(`Try 1`);
				x + 2;
			}finally{
				console.log("Finally 1");
			}

			console.log("But it's OK now");

		//▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴//
		},
		'throw1': function(){
		//▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾//
			throw 'Error 1';
			console.log("Post-throw action"); //never
		//▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴//
		},
		'throw2': function(){
		//▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾//
			try{
				throw("some error");
				console.log("Post-throw action"); //never
			}catch(e){
				console.log(typeof e);
				console.dir( e );
			}
		//▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴//
		},
		'example2': function(){
		//▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾▾//

		//▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴▴//
		},

	}

	examples.throw1();
</script>
</body>
</html>